sql/2020/03_Markup/pages_almanac_by_device_and_attribute_name_frequency.sql

```diff
@@ -0,0 +1,40 @@
+#standardSQL
+# pages almanac metrics grouped by device and element attribute use (frequency)
+
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+CREATE TEMPORARY FUNCTION get_almanac_attribute_info(almanac_string STRING)
+RETURNS ARRAY<STRUCT<name STRING, freq INT64>> LANGUAGE js AS '''
+try {
+    var almanac = JSON.parse(almanac_string);
+
+    if (Array.isArray(almanac) || typeof almanac != 'object') return [];
+
+    if (almanac.attributes_used_on_elements) {
+      return Object.entries(almanac.attributes_used_on_elements).map(([name, freq]) => ({name, freq})); 
+    }
+    
+} catch (e) {
+    
+}
+return []; 
+''';
+
+SELECT
+  _TABLE_SUFFIX AS client,
+  almanac_attribute_info.name,
+  SUM(almanac_attribute_info.freq) AS freq, # total count from all pages
+  AS_PERCENT(SUM(almanac_attribute_info.freq), SUM(SUM(almanac_attribute_info.freq)) OVER (PARTITION BY _TABLE_SUFFIX)) AS pct_m400 
+FROM
+  `httparchive.pages.2020_08_01_*`,
+  UNNEST(get_almanac_attribute_info(JSON_EXTRACT_SCALAR(payload, '$._almanac'))) AS almanac_attribute_info
+GROUP BY
+  client,
+  almanac_attribute_info.name
+ORDER BY
+  client,
+  pct_m400 DESC
+LIMIT
+  10000
```

Comment:
Consider a smaller limit. Having many rows in the spreadsheet slows it down and there may not be many interesting results down here. Alternatively you could filter using something like `HAVING pct > 0.001`.

Reply:
That came from last year. I've changed it to 1,000. I could trim it manually later if needed.

Reply:
Still showing as 10,000? The limits are kind of arbitrary but you could also use the `HAVING` suggestion.

Reply:
Looks like I missed a few. Did a search and replaced all to 1,000 or less.